internal/applause/applause.go

```diff
@@ -1 +1,278 @@
+	// `time`
+	ApplauseSince(time int64) (map[int]int, error)
+}
+
+// Applause holds the state of the service.
+type Applause struct {
+	backend   Backend
+	topic     *topic.Topic
+	datastore datastore.Getter
+}
+
+// New returns an initialized state of the notify service.
+//
+// The New function is not blocking. The context is used to stop a goroutine
+// that is started by this function.
+func New(b Backend, db datastore.Getter, closed <-chan struct{}) *Applause {
+	notify := Applause{
+		backend:   b,
+		topic:     topic.New(topic.WithClosed(closed)),
+		datastore: db,
+	}
+
+	// Make sure the topic is not empty.
+	notify.topic.Publish("")
+
+	return &notify
+}
+
+// MSG contians the current applause level and number of present users.
+type MSG struct {
+	Level        int `json:"level"`
+	PresentUsers int `json:"present_users"`
+}
+
+// Send registers, that a user applaused in a meeting.
+func (a *Applause) Send(ctx context.Context, meetingID, userID int) error {
+	if userID == 0 {
+		return iccerror.NewMessageError(iccerror.ErrNotAllowed, "Anonymous is not allowed to applause. Please be quiet.")
+	}
+
+	fetcher := datastore.NewFetcher(a.datastore)
+
+	applauseEnabled := fetcher.Field().Meeting_ApplauseEnable(ctx, meetingID)
+	if err := fetcher.Err(); err != nil {
+		return fmt.Errorf("fetching applause enabled: %w", err)
+	}
+
+	if !applauseEnabled {
+		return iccerror.NewMessageError(iccerror.ErrNotAllowed, "applause is not enabled in meeting %d. Please be quiet.", meetingID)
+	}
+
+	meetingUserIDs := fetcher.Field().Meeting_UserIDs(ctx, meetingID)
+	if err := fetcher.Err(); err != nil {
+		return fmt.Errorf("fetching meeting users: %w", err)
+	}
+
+	var inMeeting bool
+	for _, u := range meetingUserIDs {
+		if u == userID {
+			inMeeting = true
+			break
+		}
+	}
+
+	if !inMeeting {
+		return iccerror.NewMessageError(iccerror.ErrNotAllowed, "You are not part of meeting %d. Please be quiet.", meetingID)
+	}
+
+	if err := a.backend.ApplausePublish(meetingID, userID, time.Now().Unix()); err != nil {
+		return fmt.Errorf("publish applause in backend: %w", err)
+	}
+	return nil
+}
+
+// CanReceive returns an error, if the user can not receive applause.
+func (a *Applause) CanReceive(ctx context.Context, meetingID, userID int) error {
+	fetcher := datastore.NewFetcher(a.datastore)
+	if userID == 0 {
+		anonymousEnabled := fetcher.Field().Meeting_EnableAnonymous(ctx, meetingID)
+		if err := fetcher.Err(); err != nil {
+			return fmt.Errorf("fetching anonymous enabled: %w", err)
+
+		}
+		if !anonymousEnabled {
+			return iccerror.NewMessageError(iccerror.ErrNotAllowed, "Anonymous is not enabled")
+		}
+		return nil
+	}
+
+	meetingUserIDs := fetcher.Field().Meeting_UserIDs(ctx, meetingID)
+	if err := fetcher.Err(); err != nil {
+		return fmt.Errorf("fetching meeting users: %w", err)
+	}
+
+	var inMeeting bool
+	for _, u := range meetingUserIDs {
+		if u == userID {
+			inMeeting = true
+			break
+		}
+	}
```

Comment:
Is there no `find` function in go?

Reply:
If not, then extract this into an own, more generic, function

Reply:
Because then, the "outer" function is smaller and you use the exact same code even in this PR two times. This indicates bad code... very bad code...